### PR TITLE
 [ WIP ] Adding WASM libs and testing the node app and the web version

### PR DESCRIPTION
**Pull request by kalwalt**
WASM libs coming from #35 
- [x] adding WASM libs.
- [x] making the necessary modfications to import the new libs.
- [x] testing the node and web version.
- [x] testing if the generated files are not corrupted.
- [x] fix issue with float conversion.
- [x] fix issue with FS module while saving the NFT files.
- [x] ~~fixing the node app (because is not generating consistent markers)~~ It was a my mistake, it generate correct markers

### CODE REVIEW

**Comment by kalwalt** [comment]
The node app and web version fails when start to generate the FeatureSet3:

```
[info] Saving FeatureSet...
[info]   Done.
[info] Generating FeatureSet3...
[info] (2048, 1637) 220.000000[dpi]
RuntimeError: integer result unrepresentable
RuntimeError: integer result unrepresentable
/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1
(function (exports, require, module, __filename, __dirname) { var Module=typeof Module!=="undefined"?Module:{};var moduleOverrides={};var key;for(key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var arguments_=[];var thisProgram="./this.program";var quit_=function(status,toThrow){throw toThrow};var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_HAS_NODE=false;var ENVIRONMENT_IS_SHELL=false;ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_HAS_NODE=typeof process==="object"&&typeof process.versions==="object"&&typeof process.versions.node==="string";ENVIRONMENT_IS_NODE=ENVIRONMENT_HAS_NODE&&!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_WORKER;ENVIRONMENT_IS_SHELL=!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_NODE&&!ENVIRONMENT_IS_WORKER;if(Module["ENVIRONMENT"]){throw new Error("Module.ENVIRONMENT has been 

RuntimeError: abort(RuntimeError: integer result unrepresentable) at Error
    at jsStackTrace (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:23445)
    at stackTrace (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:23616)
    at process.abort (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:20292)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emitPendingUnhandledRejections (internal/process/promises.js:108:22)
    at process._tickCallback (internal/process/next_tick.js:189:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:188:16)
    at bootstrap_node.js:609:3
    at process.abort (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:20323)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emitPendingUnhandledRejections (internal/process/promises.js:108:22)
    at process._tickCallback (internal/process/next_tick.js:189:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:188:16)
    at bootstrap_node.js:609:3


```
in the browser version a similar error it is related to float to integer conversion due to Emscripten Fastcomp. We had already this kind of issue in jsarttolkit5, i will try to fix.
The problem is if we had to use the Upstream version it does not exist a Docker version with Upstream...

**Comment by kalwalt** [comment]
Testing the new libs withe the node app.js and the pinball.jpg included in jsartoolkit5:
```
[info] Generator started at 2020-02-12 12:53:44 +0100
....// omitted log
[info] Saving FeatureSet3...
[info]   Done.
[info] Generator finished at 2020-02-12 13:07:08 +0100
--
TypeError: Module.FS.readFile is not a function
TypeError: Module.FS.readFile is not a function
/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1
(function (exports, require, module, __filename, __dirname) { var Module=typeof Module!=="undefined"?Module:{};var moduleOverrides={};var key;for(key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var arguments_=[];var thisProgram="./this.program";var quit_=function(status,toThrow){throw toThrow};var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_HAS_NODE=false;var ENVIRONMENT_IS_SHELL=false;ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_HAS_NODE=typeof process==="object"&&typeof process.versions==="object"&&typeof process.versions.node==="string";ENVIRONMENT_IS_NODE=ENVIRONMENT_HAS_NODE&&!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_WORKER;ENVIRONMENT_IS_SHELL=!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_NODE&&!ENVIRONMENT_IS_WORKER;if(Module["ENVIRONMENT"]){throw new Error("Module.ENVIRONMENT has been 
RuntimeError: abort(TypeError: Module.FS.readFile is not a function) at Error
    at jsStackTrace (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:23445)
    at stackTrace (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:23616)
    at process.abort (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:20292)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emitPendingUnhandledRejections (internal/process/promises.js:108:22)
    at process._tickCallback (internal/process/next_tick.js:189:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:188:16)
    at bootstrap_node.js:609:3
    at process.abort (/home/walter/kalwalt-github/NFT-Marker-Creator/NftMarkerCreator_wasm.js:1:20323)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emitPendingUnhandledRejections (internal/process/promises.js:108:22)
    at process._tickCallback (internal/process/next_tick.js:189:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:188:16)
    at bootstrap_node.js:609:3
```

Now the float issue is solved, another issue araise with FS module. but look at the time ! i think it is faster.... but needs to be compared.

**Comment by kalwalt** [comment]
another note: maybe we should create a lib folder where to stroe the libs? Otherwise we have duplicate libs.

File: app.js [hunk 1]
```diff
@@ -4,7 +4,9 @@ const glob = require('glob');
 const inkjet = require('inkjet');
 const PNG = require('pngjs').PNG;
 const readlineSync = require('readline-sync');
-var Module = require('./NftMarkerCreator.min.js');
+//var Module = require('./NftMarkerCreator.min.js');
+var artoolkit_wasm_url = '../build/artoolkit_wasm.wasm';
```
Comment (kalwalt) [inline app.js]: Wrongs PATH and lib! i will fix this!

**Comment by kalwalt** [comment]
@Carnaux @nicolocarpignoli  with the latest the files are saved !! tested the web version, i will do a test also with the noda app. IMO less time for the creation of the files, i will confirm this when i do a real comparison ( the version in  the master and this PR)

**Comment by kalwalt** [comment]
The node app also save the files. All good, but look: it seems that the browser version is faster ??

```js
NftMarkerCreator_wasm.js:1 Generator started at 2020-02-12 15:49:44 +0100
NftMarkerCreator_wasm.js:1 [info] Generator finished at 2020-02-12 15:56:37 +0100
```
6 minutes and 53 seconds in the web version

```js
Generator started at 2020-02-12 16:06:14 +0100
[info] Generator finished at 2020-02-12 16:20:26 +0100
```

**Comment by Carnaux** [comment]
Interesting! What image did you use? Can you test with really big image?

**Comment by kalwalt** [comment]
I'm testing with the pinball [image](https://github.com/artoolkitx/jsartoolkit5/blob/master/pinball.jpg) i have found an error in the app version i will push now the fix.

**Comment by kalwalt** [comment]
with the node app: 
```
Generator started at 2020-02-12 16:28:27 +0100
[info] Generator finished at 2020-02-12 16:42:21 +0100
```
still require much time than the web version, maybe is not using WASM?

**Comment by kalwalt** [comment]
The files (.iset .fset .fset3) generated  by the web version are ok: i can track the pinball image with them in the simplest NFT example. Instead with the files generated with the node app i can not track the pinball image, they well loaded, but not track the image (also in the console there is no message of tracking !) need to understand why this. Maybe my implementation in the node app is not correct.

**Comment by kalwalt** [comment]
web version with WASM:

 ```js
Generator started at 2020-02-12 15:49:44 +0100
Generator finished at 2020-02-12 15:56:37 +0100
```
6 minutes and 53 seconds.

web wersion from master NO WASM:

```js
Generator started at 2020-02-12 17:25:12 +0100
Generator finished at 2020-02-12 17:43:25 +0100
```
18 minutes and 13 seconds.

What do you think?

**Comment by Carnaux** [comment]
That is a really significant improvement!!

**Comment by kalwalt** [comment]
> That is a really significant improvement!!

Yeah!! but please help me with the node app! and i will try also with SIMD...

**Comment by kalwalt** [comment]
And also make some tests when you have time.

**Comment by Carnaux** [comment]
@kalwalt today I have time, I will help you with the node app and make some tests

**Comment by Carnaux** [comment]
@kalwalt can you send me the NFT example you were using?

**Comment by Carnaux** [comment]
I'm testing with the nft_worker_threejs.html example in the misdake-modifies branch. And I noticed that with the original pinball NFT marker, the sphere appears in the center and with a big scale when I use the pinball marker generated by the wasm version of the node app, the sphere appears in the bottom corner and much smaller.

**Comment by kalwalt** [comment]
> I'm testing with the nft_worker_threejs.html example in the misdake-modifies branch. And I noticed that with the original pinball NFT marker, the sphere appears in the center and with a big scale when I use the pinball marker generated by the wasm version of the node app, the sphere appears in the bottom corner and much smaller.

> i didn't noticed this. But maybe depends by the dpi of the image? Did you tested the markers created with the web or node app?

**Comment by Carnaux** [comment]
The node app. It worked fine with that example

**Comment by kalwalt** [comment]
> The node app. It worked fine with that example

i have to retry maybe that file was simply corrupted. that's a good news. And what about the time for build the marker?

**Comment by Carnaux** [comment]
Generator started at 2020-02-13 16:01:35 -0200
Generator finished at 2020-02-13 16:06:42 -0200

5 min +/-

**Comment by kalwalt** [comment]
Better than mine result...! This is a great improvement in respect with the initial version!

**Comment by Carnaux** [comment]
Yes! Great job figuring that out. Can you make new tests? If you are able to track I will merge this PR. And can you make a new PR to the Dev branch with your modifications?

**Comment by kalwalt** [comment]
Tested again with the node app and the pinball.jpg.

```
Generator started at 2020-02-13 22:34:49 +0100
Generator finished at 2020-02-13 22:40:35 +0100
```
5 min. and  46 sec. not bad! Now it worked, i think before i did something wrong with the setup.
I will test with jsartoolkit5.

**Comment by Carnaux** [comment]
Nice!! Similar result to mine!

**Comment by kalwalt** [comment]
Yes!! i think before i entered wrong values for width and height....
Now is tracked, but not tested with the master jsartoolkit5 branch, i will do other tests,
PR for dev branch is already there, but wait please before merging, at least until we make all the tests. 
As i said in a previous comment that maybe should be better to put all the libs in a libs folder to avoid duplicates, what do you think in this regard?

**Comment by kalwalt** [comment]
For libs  i am refering to NftMarkerCreator.min.js ad the wasm version of this.

**Comment by kalwalt** [comment]
> I'm testing with the nft_worker_threejs.html example in the misdake-modifies branch. And I noticed that with the original pinball NFT marker, the sphere appears in the center and with a big scale when I use the pinball marker generated by the wasm version of the node app, the sphere appears in the bottom corner and much smaller.

i tested with jsartoolkit5 and master branch both markers that one coming from the web and that one coming from the node app and that one included in the repository act in the same way, i can't see this difference maybe in that branch was something that interfere?

**Comment by Carnaux** [comment]
> As i said in a previous comment that maybe should be better to put all the libs in a libs folder to avoid duplicates, what do you think in this regard? 

That is a good ideia

>  i can't see this difference maybe in that branch was something that interfere?

It might be my branch version, but if it is working is fine.

**Comment by kalwalt** [comment]
I am fine with it, you can merge both PR if you want.

**Comment by Carnaux** [comment]
Ok, now I will make some tests with the SIMD

**Pull request merged**